src/io/delimited.cr

```diff
@@ -46,6 +46,147 @@ class IO::Delimited < IO
+    else
+      read_without_peek(slice)
+    end
+  end
+
+  private def read_with_peek(slice : Bytes, peek : Bytes) : Int32
+    # If there's nothing else to peek, we reached EOF
+    if peek.empty?
+      @finished = true
+
+      if @active_delimiter_buffer.empty?
+        return 0
+      else
+        # If we have something in the active delimiter buffer,
+        # but we don't have any more data to read, that wasn't
+        # the delimiter so we must include it in the slice.
+        slice.copy_from(@active_delimiter_buffer)
+        return @active_delimiter_buffer.size
+      end
+    end
+
+    # If we have something in the active delimiter buffer
+    unless @active_delimiter_buffer.empty?
+      # This is the rest of the delimiter we have to match
+      delimiter_remaining = @read_delimiter[@active_delimiter_buffer.size..]
+
+      # This is how much we can actually match of that (peek might not have enough data!)
+      min_size = Math.min(delimiter_remaining.size, peek.size)
+
+      # See if what remains to match in the delimiter matches whatever
+      # we have in peek, limited to what's available
+      if delimiter_remaining[0, min_size] == peek[0, min_size]
+        # If peek has enough data to match the entire rest of the delimiter...
+        if peek.size >= delimiter_remaining.size
+          # We found the delimiter!
+          @io.skip(min_size)
+          @finished = true
+          return 0
+        else
+          # Copy the remaining of peek to the active delimiter buffer for now
+          (@delimiter_buffer + @active_delimiter_buffer.size).copy_from(peek)
+          @active_delimiter_buffer = @delimiter_buffer[0, @active_delimiter_buffer.size + peek.size]
+
+          # Skip whatever we had in peek, and try reading more
+          @io.skip(peek.size)
+          return read_internal(slice)
+        end
+      else
+        # No match: just copy from the active delimiter
+        read_bytes = @active_delimiter_buffer.size
+        slice.copy_from(@active_delimiter_buffer)
+        slice += read_bytes
+        @active_delimiter_buffer = Bytes.empty
+        return read_bytes + read_internal(slice)
+      end
+    end
+
+    first_byte = @read_delimiter[0]
+    index = peek.index(first_byte)
+
+    # If we can't find the delimiter's first byte we can just read from peek
+    unless index
+      # If we have more in peek than what we need to read, read all of that
+      if peek.size >= slice.size
+        slice.copy_from(peek[0, slice.size])
+        @io.skip(slice.size)
+        return slice.size
+      else
+        # Otherwise, read from peek for now
+        slice.copy_from(peek)
+        @io.skip(peek.size)
+        slice += peek.size
+        return peek.size
+      end
+    end
+
+    # If the matching byte is past the data we need to read,
+    # we can fill the entire slice
+    if index >= slice.size
+      slice.copy_from(peek[0, slice.size])
+      @io.skip(slice.size)
+      return slice.size
+    end
```

Comment:
I believe this conditional could be merged with the one above (`peek.size >= slice.size`) for more concise code.
The resulting clause structure would be: 
```cr
if (index || peek.size) >= slice.size
  # copy peek limited to slice.slice
elsif !index
  # copy entire peek
end

Reply:
I'll do it. I just merged the two comments. I think after that merge I no longer can follow the code. Yes, the code is duplicated, but I can't explain `(index || peek.size) >= slice.size` with words. Could you take a look at the comment and try to see if it could be clarified?